scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,824 @@
+                       Level.DEBUG);
+        return answer;
+    }
+
+    @Override
+    public <T extends Serializable> T remove(String originator, TaskId taskid, String channel, String key)
+            throws InvalidChannelException, IOException {
+        T answer = ((Channel<T>) getChannel(channel)).remove(key);
+        commitIfNeeded(channel);
+        logWithContext(taskid, channel, "Removed key '" + key + "', previous value was " + answer, Level.DEBUG);
+        return answer;
+    }
+
+    @Override
+    public <T extends Serializable> void putAll(String originator, TaskId taskid, String channel,
+            Map<? extends String, ? extends T> m) throws InvalidChannelException, IOException {
+        ((Channel<T>) getChannel(channel)).putAll(m);
+        logWithContext(taskid, channel, "PutAll called, with new entries: " + m, Level.DEBUG);
+        commitIfNeeded(channel);
+    }
+
+    @Override
+    public void clear(String originator, TaskId taskid, String channel) throws InvalidChannelException, IOException {
+        getChannel(channel).clear();
+        logWithContext(taskid, channel, "cleared channel", Level.DEBUG);
+        commitIfNeeded(channel);
+    }
+
+    @Override
+    public Set<String> keySet(String originator, TaskId taskid, String channel) throws InvalidChannelException {
+        return getChannel(channel).keySet();
+    }
+
+    @Override
+    public <T extends Serializable> Collection<T> values(String originator, TaskId taskid, String channel)
+            throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).values();
+    }
+
+    @Override
+    public <T extends Serializable> Set<Map.Entry<String, T>> entrySet(String originator, TaskId taskid, String channel)
+            throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).entrySet();
+    }
+
+    @Override
+    public <T extends Serializable> T getOrDefault(String originator, TaskId taskid, String channel, String key,
+            T defaultValue) throws InvalidChannelException {
+        return ((Channel<T>) getChannel(channel)).getOrDefault(key, defaultValue);
+    }
+
+    @Override
+    public <T extends Serializable> void forEach(String originator, TaskId taskid, String channel, String action)
+            throws CompilationException, ClosureEvaluationException, InvalidChannelException {
+        try {
+            Channel<T> chosenChannel = getChannel(channel);
+            chosenChannel.forEach((BiConsumer<? super String, ? super T>) evaluateClosure(action, BiConsumer.class));
+            logWithContext(taskid, channel, "Run forEach, new content is: " + chosenChannel, Level.DEBUG);
+        } catch (InvalidChannelException | CompilationException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException("Exception when evaluating closure : " +
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "Exception when evaluating closure : " 12 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)